gocourse04/SmartZoo/main.go

```diff
@@ -0,0 +1,255 @@
+			"Dining room": {
+				ID: 5,
+				Animals: []Animal{
+					{ID: 13, Name: "Hawk4"},
+				},
+			},
+			"Bath room": {
+				ID: 6,
+				Animals: []Animal{
+					{ID: 14, Name: "Cow4"},
+				},
+			},
+		},
+	},
+}
+
+// Lookup returns the found animal and in which sector and area it is.
+func (z *Zoo) Lookup(name string) (*Area, *Sector, *Animal) {
+	for _, area := range z.Areas {
+		for _, sector := range area.Sectors {
+			for _, animal := range sector.Animals {
+				if animal.Name == name {
+					return &area, &sector, &animal
+				}
+			}
+		}
+	}
+	return nil, nil, nil
+}
+
+// move the animal to another sector
+func (z *Zoo) Migration(fromAreaName, toArea string, fromSectorID, toSector int, animal Animal) error {
+	for _, area := range z.Areas {
+		if area.Name == toArea {
+			for sectorName, sector := range area.Sectors {
+				if sector.ID == toSector {
+					sector.Animals = append(sector.Animals, animal)
+					area.Sectors[sectorName] = sector // rewrite sector
+					z.deleteOldRecord(fromAreaName, fromSectorID, animal)
+					return nil
+				}
+			}
+		}
+	}
+	return fmt.Errorf("sector not found")
+}
+
+func (z *Zoo) deleteOldRecord(fromAreaName string, fromSectorID int, animal Animal) {
+	for _, area := range z.Areas {
+		if area.Name == fromAreaName {
+			for sectorName, sector := range area.Sectors {
+				if sector.ID == fromSectorID {
+					for i, a := range sector.Animals {
+						if a.ID == animal.ID {
+							sector.Animals = append(sector.Animals[:i], sector.Animals[i+1:]...)
+							area.Sectors[sectorName] = sector
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+// clean the animal. ID 6 = Bath room
+// the animal moves to the “Bath room” sector and returns to the sector from which it came.
+func (a Animal) Clean(fromArea string, sectorID int, z Zoo) {
+	var toArea string
+	var toSector int
+	for _, area := range z.Areas {
+		if area.Name == areaClean {
+			for sectorName, sector := range area.Sectors {
+				if sector.ID == sectorClean {
+					toSector = sector.ID
+					toArea = area.Name
+					sector.Animals = append(sector.Animals, a)
+					area.Sectors[sectorName] = sector // rewrite sector
+					z.deleteOldRecord(fromArea, sectorID, a)
+				}
+			}
+		}
+	}
+
+	// roll back
+	for _, area := range z.Areas {
+		if area.Name == fromArea {
+			for sectorName, sector := range area.Sectors {
+				if sector.ID == sectorID {
+					sector.Animals = append(sector.Animals, a)
+					area.Sectors[sectorName] = sector // rewrite  sector
+					z.deleteOldRecord(toArea, toSector, a)
+				}
+			}
+		}
+	}
+}
+
+// the animals go out to eat - Sector: Dining room
+func (z *Zoo) FeedAnimals(fromArea, fromSector string) error {
+	diningRoom := z.Areas["Techroom"].Sectors["Dining room"]
+	val, ok := z.Areas[fromArea].Sectors[fromSector]
+	if !ok {
+		return errors.New("sector in this Area - not found")
+	}
+
+	for _, sector := range val.Animals {
+		diningRoom.Animals = append(diningRoom.Animals, sector)
+	}
```

Comment:
Спробуй, ось так має працювати та робити те саме:
```suggestion
	diningRoom.Animals = append(diningRoom.Animals, val.Animals...)
```